piecrust/src/call_tree.rs

```diff
@@ -0,0 +1,248 @@
+    /// The zeroth previous element is the current node.
+    pub(crate) fn nth_up(&self, n: usize) -> Option<CallTreeElem> {
+        let mut current = self.0;
+
+        let mut i = 0;
+        while i < n {
+            current = current.and_then(|inner| unsafe { (*inner).prev });
+            i += 1;
+        }
+
+        current.map(|inner| unsafe { (*inner).elem })
+    }
+
+    /// Clears the call tree of all elements.
+    pub(crate) fn clear(&mut self) {
+        unsafe {
+            if let Some(inner) = self.0 {
+                let mut root = inner;
+
+                while let Some(prev) = (*root).prev {
+                    root = prev;
+                }
+
+                self.0 = None;
+                free_tree(root);
+            }
+        }
+    }
+
+    /// Returns an iterator over the call tree, starting from the rightmost
+    /// leaf, and proceeding to the top of the current position of the tree.
+    pub fn iter(&self) -> impl Iterator<Item = &CallTreeElem> {
+        CallTreeIter {
+            tree: self.0.map(|root| unsafe {
+                let mut node = root;
+
+                while !(*node).children.is_empty() {
+                    let last_index = (*node).children.len() - 1;
+                    node = (*node).children[last_index];
+                }
+
+                Subtree { root, node }
+            }),
+            _marker: PhantomData,
+        }
+    }
+}
+
+struct Subtree {
+    root: *mut CallTreeNode,
+    node: *mut CallTreeNode,
+}
+
+impl Drop for CallTree {
+    fn drop(&mut self) {
+        self.clear()
+    }
+}
+
+unsafe fn update_spent(node: *mut CallTreeNode) {
+    let node = &mut *node;
+    node.children.iter_mut().for_each(|&mut child| unsafe {
+        node.elem.spent -= (*child).elem.spent;
+        update_spent(child);
+    });
+}
+
+unsafe fn free_tree(root: *mut CallTreeNode) {
+    let mut node = Box::from_raw(root);
+
+    let mut children = Vec::new();
+    mem::swap(&mut node.children, &mut children);
+
+    for child in children {
+        free_tree(child);
+    }
+}
+
+struct CallTreeNode {
+    elem: CallTreeElem,
+    children: Vec<*mut Self>,
+    prev: Option<*mut Self>,
+}
+
+impl CallTreeNode {
+    fn new(elem: CallTreeElem) -> *mut Self {
+        Box::leak(Box::new(Self {
+            elem,
+            children: Vec::new(),
+            prev: None,
+        }))
+    }
+
+    fn with_prev(elem: CallTreeElem, prev: *mut Self) -> *mut Self {
+        Box::leak(Box::new(Self {
+            elem,
+            children: Vec::new(),
+            prev: Some(prev),
+        }))
+    }
+}
+
+/// An iterator over a [`CallTree`].
+///
+/// It starts at the righmost node and proceeds leftward towards its siblings,
+/// up toward the root.
+struct CallTreeIter<'a> {
+    tree: Option<Subtree>,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a> Iterator for CallTreeIter<'a> {
+    type Item = &'a CallTreeElem;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
```

Comment:
This is quite a long unsafe block, could it be avoided? If not, could you add a comment to it?

Reply:
Unfortunately the traversal logic is quite heavy on the pointer dereferences, so the long unsafe block is kind of unavoidable. We also have a nodes referring to both their children *and* their parent, so pointers are unavoidable as well.

I did add a comment on why all these dereferences are safe. We ensure that the tree exists - i.e. the nodes the pointers point to exist - by leveraging a lifetime into the tree when the iterator is created.